### PR TITLE
Ensure connectionSource is released in QueryBatchCursor

### PR DESCRIPTION
**Pull request by rozza**
Also ensure killCursors does not throw an exception.

JAVA-3753

https://spruce.mongodb.com/version/5eec91c70ae6064264b7ea59/tasks

### CODE REVIEW

File: driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java [hunk 1]
```diff
@@ -343,13 +338,25 @@ private void killCursor() {
...
+            } catch (MongoException e) {
+                // Ignore exceptions from calling killCursor
+            } finally {
+                serverCursor = null;
             }
-            serverCursor = null;
+        }
+    }
+
+    private void checkServerCursorAndReleaseConnectionSource() {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java:356]: Nit: consider renaming to `releaseConnectionSourceIfNoServerCursor`

File: driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java [hunk 2]
```diff
@@ -343,13 +338,25 @@ private void killCursor() {
...
-            } else {
-                connection.killCursor(namespace, singletonList(serverCursor.getId()));
+            try {
+                if (serverIsAtLeastVersionThreeDotTwo(connection.getDescription())) {
+                    connection.command(namespace.getDatabaseName(), asKillCursorsCommandDocument(), NO_OP_FIELD_NAME_VALIDATOR,
+                            ReadPreference.primary(), new BsonDocumentCodec(), connectionSource.getSessionContext());
+                } else {
+                    connection.killCursor(namespace, singletonList(serverCursor.getId()));
+                }
+            } catch (MongoException e) {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java:348]: I'm not sure if we want this behavior change of ignoring exceptions from killing cursors.  Can we defer this to a separate ticket?

**Comment by jyemin** [comment]
I'm also not sure whether it's correct to say that this is fixing the issue reported in JAVA-3753, which was really about mongos OOM errors.  How about we add a new issue just about properly releasing connection source and link it?

File: driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java (hunk 1 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java]: Done

File: driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java (hunk 2 above)
Comment (rozza) [reply driver-core/src/main/com/mongodb/internal/operation/QueryBatchCursor.java]: Reverted

**Comment by rozza** [comment]
Added JAVA-3774 for this change.

**Pull request closed without merging**